DuckDuckGo/Feedback/VPNMetadataCollector.swift

```diff
@@ -0,0 +1,240 @@
+            return encodedMetadata.base64EncodedString()
+        } catch {
+            return "Failed to encode metadata to JSON, error message: \(error.localizedDescription)"
+        }
+    }
+}
+
+protocol VPNMetadataCollector {
+    func collectMetadata() async -> VPNMetadata
+}
+
+final class DefaultVPNMetadataCollector: VPNMetadataCollector {
+    private let statusObserver: ConnectionStatusObserver
+    private let serverInfoObserver: ConnectionServerInfoObserver
+
+    init(statusObserver: ConnectionStatusObserver = ConnectionStatusObserverThroughSession(),
+         serverInfoObserver: ConnectionServerInfoObserver = ConnectionServerInfoObserverThroughSession()) {
+        self.statusObserver = statusObserver
+        self.serverInfoObserver = serverInfoObserver
+    }
+
+    func collectMetadata() async -> VPNMetadata {
+        let appInfoMetadata = collectAppInfoMetadata()
+        let deviceInfoMetadata = collectDeviceInfoMetadata()
+        let networkInfoMetadata = await collectNetworkInformation()
+        let vpnState = await collectVPNState()
+        let vpnSettingsState = collectVPNSettingsState()
+
+        return VPNMetadata(
+            appInfo: appInfoMetadata,
+            deviceInfo: deviceInfoMetadata,
+            networkInfo: networkInfoMetadata,
+            vpnState: vpnState,
+            vpnSettingsState: vpnSettingsState
+        )
+    }
+
+    // MARK: - Metadata Collection
+
+    private func collectAppInfoMetadata() -> VPNMetadata.AppInfo {
+        let appVersion = AppVersion.shared.versionNumber
+        let versionStore = NetworkProtectionLastVersionRunStore()
+        let isInternalUser = AppDependencyProvider.shared.internalUserDecider.isInternalUser
+
+        return .init(appVersion: appVersion, lastVersionRun: versionStore.lastVersionRun ?? "Unknown", isInternalUser: isInternalUser)
+    }
+    
+    private func collectDeviceInfoMetadata() -> VPNMetadata.DeviceInfo {
+        .init(osVersion: AppVersion.shared.osVersion, lowPowerModeEnabled: ProcessInfo.processInfo.isLowPowerModeEnabled)
+    }
+    
+    func collectNetworkInformation() async -> VPNMetadata.NetworkInfo {
+        let monitor = NWPathMonitor()
+        monitor.start(queue: DispatchQueue(label: "VPNMetadataCollector.NWPathMonitor.paths"))
+
+        var path: Network.NWPath?
+        let startTime = CFAbsoluteTimeGetCurrent()
+
+        let dateFormatter = DateFormatter()
+        dateFormatter.calendar = Calendar.current
+        dateFormatter.timeZone = TimeZone(secondsFromGMT: 0)
+        dateFormatter.dateFormat = "yyyy-MM-dd HH:mm:ss"
+
+        let networkPathChange = VPNSettings(defaults: .networkProtectionGroupDefaults).networkPathChange
+        let now = Date()
+
+        while true {
+            if !monitor.currentPath.availableInterfaces.isEmpty {
+                path = monitor.currentPath
+                monitor.cancel()
+
+                let changeDate = networkPathChange?.date ?? .distantPast
```

Comment:
It's possible to get `.distantPast` if you submit feedback right after starting NetP for the first time and not yet changing networks - I could see the setup flow being a source of feedback, what do you think about using some other value here?

I was thinking that `secondsSincePathChange` could become be converted from int to string, and in the result of no path change then it can just say `never changed` or something.

Reply:
Makes sense. I’d change that to report “unknown” if changeDate is nil 👍